.azure-pipelines.yml

```diff
@@ -0,0 +1,248 @@
+name: $(Build.SourceBranch)-$(date:yyyyMMdd)$(rev:.r)
+trigger:
+  branches:
+    include:
+    - 'master'
+  tags:
+    include:
+    - '*'
+
+jobs:
+- job: Build
+  strategy:
+    matrix:
+      windows-stable:
+        imageName: 'vs2017-win2016'
+        rustup_toolchain: stable
+      windows-stable-release:
+        imageName: 'vs2017-win2016'
+        rustup_toolchain: stable
+        build_type: release
+      linux-stable-release:
+        imageName: 'ubuntu-16.04'
+        rustup_toolchain: stable
+        build_type: release
+      mac-stable-release:
+        imageName: 'macos-10.14'
+        rustup_toolchain: stable
+        build_type: release
+      linux-stable:
+        imageName: 'ubuntu-16.04'
+        rustup_toolchain: stable
+      mac-stable:
+        imageName: 'macos-10.14'
+        rustup_toolchain: stable
+      mac-beta:
+        imageName: 'macos-10.14'
+        rustup_toolchain: beta
+      mac-nightly:
+        imageName: 'macos-10.14'
+        rustup_toolchain: nightly
+
+  pool:
+    vmImage: $(imageName)
+
+  variables:
+    SCCACHE_DIR: $(Pipeline.Workspace)/.sccache
+    RUSTC_WRAPPER: sccache
+
+  steps:
+    - checkout: self
+      submodules: true
+    - script: |
+        curl https://sh.rustup.rs -sSf | sh -s -- -y --default-toolchain $RUSTUP_TOOLCHAIN
+        echo "##vso[task.setvariable variable=PATH;]$PATH:$HOME/.cargo/bin"
+        export PATH=$PATH:$HOME/.cargo/bin
+        rustup update $RUSTUP_TOOLCHAIN
+        rustup default $RUSTUP_TOOLCHAIN
+        rustc --version
+      displayName: Install rust (*nix)
+      condition: ne( variables['Agent.OS'], 'Windows_NT' )
+    - script: |
+        rustup update %RUSTUP_TOOLCHAIN%
+        rustup default %RUSTUP_TOOLCHAIN%
+        rustc --version
+      displayName: Install rust (Win)
+      condition: eq( variables['Agent.OS'], 'Windows_NT' )
+    - script: |
+        rustup component add rustfmt
+      displayName: Add rustfmt
+      condition: and(eq(variables['rustup_toolchain'], 'stable'), ne( variables['build_type'], 'release' ),
+                     eq(variables['Agent.OS'], 'Darwin'))
```

Comment:
This is probably fine to just go ahead and execute everywhere since it's so cheap